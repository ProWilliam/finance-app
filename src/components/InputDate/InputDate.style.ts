import { StyleSheet } from 'react-native';
import theme from '../../theme/theme';

const StyleInputDate = StyleSheet.create({
  container: {
    alignItems: 'center',
    backgroundColor: theme.light.tertiary,
    borderRadius: 5,
    color: theme.light.primary
  },
  textInput: {
    color: theme.light.quaternary,
    fontSize: theme.size.fontLittle,
    padding: 10,
  },
  inputWeb: {
    height: 50,
    paddingLeft: 20,
    paddingRight: 20,
    borderColor: theme.light.border,
    borderWidth: 1,
    borderRadius: 5,
  }
});


export default StyleInputDate