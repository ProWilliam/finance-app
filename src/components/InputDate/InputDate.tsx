import { useState } from 'react';
import { View, Platform, Text, TouchableOpacity } from 'react-native';
import DateTimePickerModal from "react-native-modal-datetime-picker";
import { FormatingDate } from '../../utils/FormatingDate';

// Type and Styles.
import styles from './InputDate.style';
import DateProps from '../../types/components/InputDate.types';



const InputDate: React.FC<DateProps> = ({ title, keyValue, onChangeForm, value }) => {

  const [dateWeb, setDateWeb] = useState('2026-08-22')
  
  const [isDatePickerVisible, setDatePickerVisibility] = useState(false);
  const [buttonTitle, setButtonTitle] = useState<string>(title);

  const showDatePicker = () => {
    setDatePickerVisibility(true);
  };

  const hideDatePicker = () => {
    setDatePickerVisibility(false);
  };

  const onDateWeb = (event: any) => {
    const dateFormWeb = event.target.value;

    onChangeForm(keyValue, dateFormWeb);
    setDateWeb(dateFormWeb)
  }

  const handleConfirm = (date: any) => {
    const dateFormat = FormatingDate(date);

    onChangeForm(keyValue, dateFormat);
    setButtonTitle(dateFormat);
    hideDatePicker();
  };

  return (
    <View>
      { Platform.OS == 'web' && (
        <input
          style={styles.inputWeb}
          type="date"
          value={dateWeb}
          onChange={onDateWeb}
        />
      )}
      { Platform.OS != 'web' && (
        <View>
          <TouchableOpacity
            style={styles.container}
            onPress={showDatePicker}
          >
            <Text
              style={styles.textInput}
            >
              {value || buttonTitle}
            </Text>
          </TouchableOpacity>
          <DateTimePickerModal
            isVisible={isDatePickerVisible}
            mode="date"
            onConfirm={handleConfirm}
            onCancel={hideDatePicker}
          />
        </View>
        
      )}
    </View>
  );
};

export default InputDate;