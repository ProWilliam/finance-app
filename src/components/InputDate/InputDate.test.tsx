import React from 'react';
import renderer, { act } from 'react-test-renderer';
import { View, Platform, Text, TouchableOpacity } from 'react-native';

import DateTimePickerModal from "react-native-modal-datetime-picker";
import InputDate from './InputDate';

// The third party library has different events for the same value, which is why any is placed as the type.
jest.mock('react-native-modal-datetime-picker', () => {
  return ({ isVisible, onConfirm, onCancel }: any) => {
    return null;
  };
});

describe('<InputDate />', () => {

  const onChangeMock = jest.fn();

  it('renders correctly with default title', () => {
    const component = renderer.create(
      <InputDate
        title="Select Date"
        keyValue="date_release"
        onChangeForm={onChangeMock}
        value=""
      />
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render a TouchableOpacity and DateTimePickerModal on non-web platforms', () => {
    Platform.OS = 'ios';
    const tree = renderer.create(<InputDate title="Fecha" keyValue="date_release" onChangeForm={onChangeMock} value="" />);

    const touchableOpacity = tree.root.findByType(TouchableOpacity);
    const dateTimePickerModal = tree.root.findByType(DateTimePickerModal);

    expect(touchableOpacity).toBeTruthy();
    expect(dateTimePickerModal.props.isVisible).toBe(false);
  });

  it('should render an input field on web platform', () => {
    Platform.OS = 'web'; // Mock the platform to web

    const component = renderer.create(
      <InputDate title="fecha" keyValue="date_release" onChangeForm={onChangeMock} value="2024-08-10" />
    );

    const input = component.root.findByType('input');
    expect(input.props.value).toBe('2024-08-10');
  });
  
  it('should update button title and form value when date is confirmed', () => {
    Platform.OS = 'ios';
    
    const tree = renderer.create(<InputDate title="Fecha" keyValue="date_release" onChangeForm={onChangeMock} value="" />);
    const instance = tree.root;

    // Simulates tapping on the button to show the DatePicker.
    const button = instance.findByType(TouchableOpacity);
    act(() => {
      button.props.onPress();
    });

    // Simulates the confirmation of a date.
    const datePicker = instance.findByType(DateTimePickerModal);
    act(() => {
      datePicker.props.onConfirm(new Date('8/10/2024'));
    });
  
    expect(onChangeMock).toHaveBeenCalledWith('date_release', '2024-08-10');
    expect(tree.root.findByType(Text).props.children).toBe('2024-08-10');
  });

  it('should render TouchableOpacity and Text on native platform', () => {
    Platform.OS = 'ios'; // Mock the platform to iOS

    const component = renderer.create(
      <InputDate title="fecha" keyValue="date_revision" onChangeForm={onChangeMock} value="2024-08-10" />
    );

    const touchableOpacity = component.root.findByType(TouchableOpacity);
    const text = component.root.findByType(Text);

    expect(touchableOpacity).toBeDefined();
    expect(text.props.children).toBe('2024-08-10');
  });

  it('should call onChangeForm with the correct date format when a date is confirmed', () => {
    const mockOnChangeForm = jest.fn();
    const component = renderer.create(
      <InputDate title="Date" keyValue="date_revision" onChangeForm={mockOnChangeForm} value="" />
    );

    const instance = component.root;
    const button = instance.findByType(TouchableOpacity);
    const datepiker = instance.findByType(DateTimePickerModal);

    // Mock the onConfirm callback
    act(() => {
      button.props.onPress();
    });

    act(() => {
      datepiker.props.onConfirm(new Date('8/10/2024'));
    });

    expect(mockOnChangeForm).toHaveBeenCalledWith('date_revision', '2024-08-10');
    expect(instance.findByType(Text).props.children).toBe('2024-08-10');
  });

  

});