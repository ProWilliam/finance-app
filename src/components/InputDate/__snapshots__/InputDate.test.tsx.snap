// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<InputDate /> renders correctly with default title 1`] = `
<View>
  <View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "backgroundColor": "#eaeaea",
          "borderRadius": 5,
          "color": "#FFFFFF",
          "opacity": 1,
        }
      }
    >
      <Text
        style={
          {
            "color": "#1E2126",
            "fontSize": 24,
            "padding": 10,
          }
        }
      >
        Select Date
      </Text>
    </View>
  </View>
</View>
`;
